src/main/java/eu/dissco/core/translator/repository/DwcaRepository.java

```diff
@@ -28,48 +37,112 @@
+    fieldsToCheck.add(0, uniqueField);
+    for (var fieldName : fieldsToCheck) {
+      if (dbRecord.getRight().has(fieldName) && !dbRecord.getRight().get(fieldName).isNull()) {
+        return dbRecord.getRight().get(fieldName).asText();
+      }
+    }
+    log.warn("No physicalSpecimenID found for record: {}", dbRecord.getLeft());
+    throw new IllegalDataException(
+        "No physicalSpecimenID found for record: " + dbRecord.getRight());
+  }
+
+  public void createTable(String tableName, Term fileType) {
     context.createTable(tableName)
-        .column(idField, SQLDataType.VARCHAR)
+        .column(dwcaIDField, SQLDataType.VARCHAR)
+        .column(uniqueIDField, SQLDataType.VARCHAR)
         .column(dataField, SQLDataType.JSONB)
         .execute();
-    context.createIndex().on(tableName, idField.getName()).execute();
+    if (fileType.equals(DwcTerm.Occurrence)) {
+      context.createUniqueIndex().on(tableName, uniqueIDField.getName()).execute();
+    } else if (fileType.prefixedName().equals("http://rs.tdwg.org/ac/terms/Multimedia")
+        || fileType.equals(GbifTerm.Multimedia)) {
+      context.createUniqueIndex().on(tableName, dwcaIDField.getName(), uniqueIDField.getName())
+          .execute();
+    }
+    context.createIndex().on(tableName, dwcaIDField.getName()).execute();
   }
 
   private Table<Record> getTable(String tableName) {
     return DSL.table("\"" + tableName + "\"");
   }
 
-  public void postRecords(String tableName, List<Pair<String, JsonNode>> dbRecords)
-      throws DisscoRepositoryException {
-    batchInserter.batchCopy(tableName, dbRecords);
+  public void postRecords(String tableName, Term fileType, List<Pair<String, JsonNode>> dbRecords) {
+    var queries = dbRecords.stream().map(dbRecord -> recordToQuery(tableName, fileType, dbRecord))
+        .filter(
+            Objects::nonNull).toList();
+    context.batch(queries).execute();
+  }
+
+  private Query recordToQuery(String tableName, Term fileType, Pair<String, JsonNode> dbRecord) {
+    try {
+      var query = context.insertInto(getTable(tableName)).set(dwcaIDField, dbRecord.getLeft())
+          .set(dataField,
+              JSONB.jsonb(mapper.writeValueAsString(dbRecord.getRight()).replace("\\u0000", "")));
+      if (fileType.equals(DwcTerm.Occurrence)) {
+        var uniqueField = dataMappingComponent.getFieldMappings()
+            .getOrDefault("ods:physicalSpecimenID", "dwc:occurrenceID");
+        return query.set(uniqueIDField, getPhysicalSpecimenID(dbRecord, uniqueField));
+      }
+      if (fileType.prefixedName().equals("http://rs.tdwg.org/ac/terms/Multimedia")
+          || fileType.equals(GbifTerm.Multimedia)) {
+        var uniqueField = dataMappingComponent.getFieldMappings()
+            .getOrDefault("ac:accessURI", "ac:accessURI");
+        return query.set(uniqueIDField, getAccessURI(dbRecord, uniqueField))
+            .onConflict(dwcaIDField, uniqueIDField)
```

Comment:
I guess it's not much of a unique id field huh... :smiling_face_with_tear:

Reply:
True, but yeah unique within the specimen.